libs/fairness/src/lib/util/calculateFairnessMetric.ts

```diff
@@ -0,0 +1,179 @@
+  }
+
+  if (fairnessMethod === FairnessModes.Max) {
+    response = { overall: max };
+    if (maxLowerBound && maxUpperBound) {
+      response = {
+        ...response,
+        bounds: {
+          lower: maxLowerBound,
+          upper: maxUpperBound
+        }
+      };
+    }
+    return response;
+  }
+
+  if (fairnessMethod === FairnessModes.Ratio) {
+    if (binBounds && binBounds.length > 1) {
+      let minRatio = Infinity;
+      let maxRatio = -Infinity;
+      for (let i = 0; i < binBounds.length - 1; i++) {
+        for (let j = i + 1; j < binBounds.length; j++) {
+          let minCandidate;
+          let maxCandidate;
+
+          // if there is overlap
+          if (
+            binBounds[i].upper > binBounds[j].lower &&
+            binBounds[j].upper > binBounds[i].lower
+          ) {
+            const minCandidate1 = binBounds[i].lower / binBounds[j].upper;
+            const minCandidate2 = binBounds[j].lower / binBounds[i].upper;
+            const minCandidate =
+              minCandidate1 < minCandidate2 ? minCandidate1 : minCandidate2;
+
+            minRatio = minCandidate < minRatio ? minCandidate : minRatio;
+            maxRatio = 1;
+          } else {
+            // index i is completely greater than j
+            if (binBounds[i].upper > binBounds[j].upper) {
+              minCandidate = binBounds[j].lower / binBounds[i].upper;
+              maxCandidate = binBounds[j].upper / binBounds[i].lower;
+            } else {
+              minCandidate = binBounds[i].lower / binBounds[j].upper;
+              maxCandidate = binBounds[i].upper / binBounds[j].lower;
+            }
+
+            minRatio = minCandidate < minRatio ? minCandidate : minRatio;
+            maxRatio = maxCandidate > maxRatio ? maxCandidate : maxRatio;
+          }
+        }
+      }
+      return {
+        bounds: {
+          lower: minRatio,
+          upper: maxRatio
+        },
+        overall: min / max
+      };
+    }
+    return {
+      overall: min / max
+    };
+  } else if (fairnessMethod === FairnessModes.Difference) {
+    if (binBounds && binBounds.length > 1) {
+      let minDiff = Infinity;
+      let maxDiff = -Infinity;
+      for (let i = 0; i < binBounds.length - 1; i++) {
+        for (let j = i + 1; j < binBounds.length; j++) {
+          let minCandidate;
+          let maxCandidate;
+
+          // if there is overlap
+          if (
+            binBounds[i].upper > binBounds[j].lower &&
+            binBounds[j].upper > binBounds[i].lower
+          ) {
+            const maxCandidate1 = Math.abs(
+              binBounds[j].upper - binBounds[i].lower
+            );
+            const maxCandidate2 = Math.abs(
+              binBounds[i].upper - binBounds[j].lower
+            );
+            maxCandidate =
+              maxCandidate1 > maxCandidate2 ? maxCandidate1 : maxCandidate2;
+
+            minDiff = 0;
+            maxDiff = maxCandidate > maxDiff ? maxCandidate : maxDiff;
```

Comment:
A comment.... doesn't doing things this way mean that the bound which contribute to the bound on the difference might not be the point which contributes to the maximum difference itself?

I'm thinking of a case where you have:
- 0.2 +/- 0.1
- 0.8 +/- 0.1
- 0.5 +/ 0.45

The difference will be 0.8-0.2=0.6, but the difference bound will be dominated by the bound on 0.5 (if I'm thinking things through correctly).

I think that your behaviour is reasonable, but it should be documented - both with a comment here and a test case or two below (with their own comments, highlighting what's going on).

Reply:
I think this behavior naturally follows the technique of coming up with the most conservative bounds. What do you think should be specifically commented on? I don't think there is anything special from the fact that the bounds on 0.5 contributed to the upper and lower bounds of the difference, but 0.5 did not contribute to the single point/aggregate difference

Reply:
Just a note that when computing the bounds on the difference, all bounds are considered, not just those at the nominal extremes. I agree that this is an entirely reasonable approach.